lib/VirtualTimeController.js

```diff
@@ -0,0 +1,353 @@
+    this._inNotifyTasksAndAdvance = false;
+    return;
+  }
+
+  this._setVirtualTimePolicy(nextBudget, false /* wait_for_navigation */);
+  this._inNotifyTasksAndAdvance = false;
+};
+
+/**
+ * @param {!VirtualTimeController.RepeatingTask} repeatingTask
+ * @param {!VirtualTimeController.ContinuePolicy} continuePolicy
+ */
+VirtualTimeController.prototype._taskReadyToAdvance = function(repeatingTask, continuePolicy) {
+  repeatingTask.readyToAdvance = true;
+  repeatingTask.continuePolicy = continuePolicy;
+  this._notifyTasksAndAdvance();
+};
+
+/**
+ * @param {number} budget
+ * @param {boolean} waitForNavigation
+ */
+VirtualTimeController.prototype._setVirtualTimePolicy = function(budget, waitForNavigation) {
+  this._lastBudget = budget;
+  const controller = this;
+  this._client.send('Emulation.setVirtualTimePolicy', {policy: 'pauseIfNetworkFetchesPending', budget, waitForNavigation})
+      .then(function(result) {
+        if (this._shouldSendStartNotification) {
+          this._shouldSendStartNotification = false;
+          const totalElapsedTimeOffset = this._totalElapsedTimeOffset;
+          this._observers.forEach(observer => observer.virtualTimeStarted(totalElapsedTimeOffset));
+        }
+      }.bind(controller));
+};
+
+/**
+ * @param {*} event
+ */
+VirtualTimeController.prototype._virtualTimeBudgetExpired = function(event) {
+  this._totalElapsedTimeOffset += this._lastBudget;
+  this._virtualTimePaused = true;
+  this._notifyTasksAndAdvance();
+};
+
+/**
+ * @param {number} budget The number of virtual milliseconds of virtual time budget to grant.
+ * @param {VirtualTimeController.StartPolicy} startPolicy Whether or not virtual time should start
+ *    advancing immediately or wait for a navigation.
+ */
+VirtualTimeController.prototype.grantVirtualTimeBudget = function(budget, startPolicy) {
+  const controller = this;
+  return new Promise(resolve => {
+    new class {
+      constructor() {
+        this._intervalElapsed = false;
+        controller.addObserver(this);
+        controller.scheduleRepeatingTask(this, budget, 0, startPolicy);
+        controller.startVirtualTime();
+      }
+
+      /**
+       * @param {number} virtualTimeOffset
+       * @param {!function(!VirtualTimeController.ContinuePolicy)} continueCallback
+       */
+      intervalElapsed(virtualTimeOffset, continueCallback) {
+        this._intervalElapsed = true;
+        continueCallback(VirtualTimeController.ContinuePolicy.STOP);
+      }
+
+      /** @param {number} virtualTimeOffset */
+      virtualTimeStarted(virtualTimeOffset) {
+      }
+
+      /** @param {number} virtualTimeOffset */
+      virtualTimeStopped(virtualTimeOffset) {
+        if (this._intervalElapsed) {
+          controller.removeObserver(this);
+          controller.cancelRepeatingTask(this);
+          resolve();
+        }
+      }
+    }();
+  });
+};
+
+
+module.exports = VirtualTimeController;
+helper.tracePublicAPI(VirtualTimeController);
```

Comment:
`VirtualTimeController` is not exposed to the users, isn't it? In this case, the `helper.tracePublicAPI()` call is not needed.

Reply:
Done